pydantic/types.py

```diff
@@ -687,6 +689,110 @@ def get_secret_value(self) -> bytes:
         return self._secret_value
 
 
+class PaymentCardBrand(Enum):
+    amex = 'American Express'
+    mastercard = 'Mastercard'
+    visa = 'Visa'
+    other = 'other'
+
+    def __str__(self) -> str:
+        return self.value
+
+
+class PaymentCardNumber(str):
+    """
+    Based on: https://en.wikipedia.org/wiki/Payment_card_number
+    """
+
+    strip_whitespace = True
+    min_length: int = 12
+    max_length: int = 19
+    bin: str
+    last4: str
+    brand: PaymentCardBrand
+
+    def __new__(cls, card_number: str) -> 'PaymentCardNumber':
+        obj = str.__new__(cls, card_number)  # type: ignore
```

Comment:
mypy didn't like this: `pydantic\types.py:715: error: Too many arguments for "__new__" of "object"`

but, it works without any issues:
```
In [1]: str.__new__?
Signature: str.__new__(*args, **kwargs)
Docstring: Create and return a new object.  See help(type) for accurate signature.
Type:      builtin_function_or_method
```

Reply:
Shouldn’t it just be `str.__new__(card_number)`?

Reply:
I originally thought that too, but ...

```
In [5]: class S(str):
   ...:     def __new__(cls, v):
   ...:         obj = str.__new__(v)
   ...:         obj.hello = 'world'
   ...:         return obj
   ...:

In [6]: s = S('dd')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-f1f5df8a6c30> in <module>
----> 1 s = S('dd')

<ipython-input-5-2531cfdd5e3a> in __new__(cls, v)
      1 class S(str):
      2     def __new__(cls, v):
----> 3         obj = str.__new__(v)
      4         obj.hello = 'world'
      5         return obj

TypeError: str.__new__(X): X is not a type object (str)
```

Reply:
from the cpython tests: https://github.com/python/cpython/blob/3.7/Lib/test/test_super.py#L153-L157

```python
        class Meta(type):
            def __new__(cls, name, bases, namespace):
                nonlocal test_class
                self = super().__new__(cls, name, bases, namespace)
                test_class = self.f()
                return self
```

I'll update to use `self` vs `obj` and `super()` vs `str` to stay consistent.

Reply:
@matin you are right about needing to pass the class to `str.__new__` -- it looks like `str.__new__` may be a `staticmethod` rather than a `classmethod` weirdly. But the example you are showing above is for a metaclass -- I don't believe `str` is a metaclass, so its `__new__` won't have that signature. I could be wrong though; I'm going to look into this more..

Reply:
I played with this a little more, and you don't need to use `__new__` at all. In fact, if I understand correctly, you probably should *not* use `__new__` like this -- your current implementation will return an actual string *instead* of an instance of a PaymentCardNumber when you call `PaymentCardNumber(card_number)`.

If you instead just create an `__init__` method, I think it would be more natural:
```python
class PaymentCardNumber(str):
    """
    Based on: https://en.wikipedia.org/wiki/Payment_card_number
    """

    strip_whitespace = True
    min_length: int = 12
    max_length: int = 19
    bin: str
    last4: str
    brand: PaymentCardBrand

    def __init__(self, card_number):
        super().__init__()  # I think this can be removed safely
        self.bin = card_number[:6]
        self.last4 = card_number[-4:]
        self.brand = PaymentCardNumber.get_brand(card_number)
```

Is there a downside to this approach that I'm missing?

Reply:
(Note: I edited the above comment a couple times to remove incorrect information -- sorry for any confusion!)

Reply:
I could _swear_ that I tried that! :-/

Regardless, this actually makes more sense to ensure those fields don't get out of date: eb3bb2a7da5ea323c57ec328220ee40e51df1844

Reply:
Actually, since `str` is immutable, do you need to worry about the fields getting out of date? Given that, it would seem to me like the main tradeoff is memory usage vs performance; either way the impact will be pretty small.

Reply:
fair point!

Reply:
cython didn't like the approach with properties: https://travis-ci.org/samuelcolvin/pydantic/jobs/579543860

The approach with `__init__` doesn't appear to work either:

```python
    def __init__(self, card_number):
>       self.bin = card_number[:6]
E       AttributeError: can't set attribute

pydantic/types.py:716: AttributeError
```

even if I leave `super().__init()` in
```python
    def __init__(self, card_number):
        super().__init__()
>       self.bin = card_number[:6]
E       AttributeError: can't set attribute

pydantic/types.py:716: AttributeError
```

reverting back to `__new__`